File: backend/app/schemas/customers.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.invoices import InvoiceResponse


class CustomerCreate(BaseModel):
    name: str = Field(min_length=1)
    customer_code: str | None = None
    company_name: str | None = None
    email: str | None = None
    phone: str | None = None
    address: str | None = None
    billing_address: str | None = None
    shipping_address: str | None = None
    status: str = "ACTIVE"
    notes: str | None = None
    credit_limit: Decimal = Field(default=0, ge=0)
    discount_percent: Decimal = Field(default=0, ge=0, le=100)


class CustomerResponse(BaseModel):
    id: UUID
    organization_id: UUID
    customer_code: str | None
    name: str
    company_name: str | None
    email: str | None
    phone: str | None
    address: str | None
    billing_address: str | None
    shipping_address: str | None
    status: str
    notes: str | None
    credit_limit: Decimal
    discount_percent: Decimal
    user_id: UUID | None
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CustomerUpdate(BaseModel):
    name: str | None = None
    customer_code: str | None = None
    company_name: str | None = None
    email: str | None = None
    phone: str | None = None
    address: str | None = None
    billing_address: str | None = None
    shipping_address: str | None = None
    status: str | None = None
    notes: str | None = None
    credit_limit: Decimal | None = Field(default=None, ge=0)
    discount_percent: Decimal | None = Field(default=None, ge=0, le=100)
    is_active: bool | None = None

    model_config = {"from_attributes": True}


class CustomerAccountCreate(BaseModel):
    email: str
    password: str = Field(min_length=8)


class CustomerContactCreate(BaseModel):
    name: str = Field(min_length=1)
    email: str | None = None
    phone: str | None = None
    job_title: str | None = None
    is_primary: bool = False
    is_active: bool = True


class CustomerContactResponse(CustomerContactCreate):
    id: UUID
    organization_id: UUID
    customer_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CustomerAddressCreate(BaseModel):
    address_type: str = Field(pattern="^(BILLING|SHIPPING|OTHER)$")
    label: str | None = None
    line1: str = Field(min_length=1)
    line2: str | None = None
    city: str | None = None
    state: str | None = None
    postal_code: str | None = None
    country: str | None = None
    is_primary: bool = False
    is_active: bool = True


class CustomerAddressResponse(CustomerAddressCreate):
    id: UUID
    organization_id: UUID
    customer_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CustomerPage(BaseModel):
    items: list[CustomerResponse]
    page: int
    page_size: int
    total: int
    total_pages: int
    has_next: bool
    has_previous: bool


class CustomerPaymentResponse(BaseModel):
    id: UUID
    order_id: UUID
    order_number: str
    invoice_number: str | None
    amount: Decimal
    payment_method: str
    reference: str | None
    status: str
    created_at: datetime

    model_config = {"from_attributes": True}


class CustomerOrderPage(BaseModel):
    items: list
    page: int
    page_size: int
    total: int
    total_pages: int
    has_next: bool
    has_previous: bool


class CustomerInvoicePage(CustomerOrderPage):
    pass


class CustomerPaymentPage(CustomerOrderPage):
    items: list[CustomerPaymentResponse]


class CustomerContactPage(CustomerOrderPage):
    items: list[CustomerContactResponse]


class CustomerAddressPage(CustomerOrderPage):
    items: list[CustomerAddressResponse]


class CustomerInvoicePortalResponse(BaseModel):
    invoice: InvoiceResponse
    payments: list[CustomerPaymentResponse]

File: frontend/types/customers.ts
export type Customer = {
  id: string;
  organization_id: string;
  customer_code: string | null;
  name: string;
  company_name: string | null;
  email: string | null;
  phone: string | null;
  address: string | null;
  billing_address: string | null;
  shipping_address: string | null;
  status: string;
  notes: string | null;
  credit_limit: string;
  discount_percent: string;
  user_id: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string;
};

export type CustomerCreate = {
  name: string;
  customer_code?: string;
  company_name?: string;
  email?: string;
  phone?: string;
  address?: string;
  billing_address?: string;
  shipping_address?: string;
  status?: string;
  notes?: string;
  credit_limit?: string;
  discount_percent?: string;
};

export type CustomerUpdate = {
  name?: string;
  customer_code?: string;
  company_name?: string;
  email?: string;
  phone?: string;
  address?: string;
  billing_address?: string;
  shipping_address?: string;
  status?: string;
  notes?: string;
  credit_limit?: string;
  discount_percent?: string;
  is_active?: boolean;
};

export type CustomerSummary = {
  customer_id: string;
  order_count: number;
  invoice_count: number;
  payment_count: number;
  sales_total: string;
  invoiced_total: string;
  paid_total: string;
  outstanding_balance: string;
  overdue_amount: string;
};

export type CustomerPage = {
  items: Customer[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
  has_next: boolean;
  has_previous: boolean;
};

export type CustomerPortalData = {
  customer: Customer;
  summary: CustomerSummary;
  contacts: CustomerContact[];
  addresses: CustomerAddress[];
  orders: Array<{ id: string; order_number: string; status: string; payment_status: string; total: string; created_at: string }>;
  invoices: Array<{ id: string; invoice_number: string; status: string; total: string; amount_paid: string; due_date: string | null; created_at: string }>;
  payments: CustomerPayment[];
};

export type CustomerContact = { id: string; customer_id: string; name: string; email: string | null; phone: string | null; job_title: string | null; is_primary: boolean; is_active: boolean; created_at: string; updated_at: string };
export type CustomerAddress = { id: string; customer_id: string; address_type: string; label: string | null; line1: string; line2: string | null; city: string | null; state: string | null; postal_code: string | null; country: string | null; is_primary: boolean; is_active: boolean; created_at: string; updated_at: string };
export type CustomerPayment = { id: string; order_id: string; order_number: string; invoice_number: string | null; amount: string; payment_method: string; reference: string | null; status: string; created_at: string; updated_at: string };
export type CustomerOrderDetail = CustomerPortalData["orders"][number] & { branch_id: string; customer_id: string | null; subtotal: string; discount: string; tax: string; notes: string | null; items: Array<{ id: string; product_id: string; product_name?: string | null; product_sku?: string | null; quantity: string; fulfilled_quantity?: string; unit_price: string; discount: string; tax: string; line_total: string }> };
export type CustomerInvoiceDetail = CustomerPortalData["invoices"][number] & { order_id: string; customer_id: string | null; issued_date: string | null; subtotal: string; discount: string; tax: string; line_items: Array<{ id: string; product_name: string; quantity: string; line_total: string }> };
export type CustomerInvoicePortalDetail = { invoice: CustomerInvoiceDetail; payments: CustomerPayment[] };
